documentation/specs/proposed/BuildCheck-Rules-Identification.md

```diff
@@ -0,0 +1,113 @@
+# MSBuild Rules/Analyzers Identification
+
+## Background and Context
+
+The MSBuild team is currently working on delivering the MSBuild analyzers (aka BuildCheck). The BuildCheck infrastructure has built-in analyzers and functionality to support custom ones. Hence, we need to make sure it will be possible to configure and differentiate built-in and custom analyzers.
+
+Note: Single analyzer can have multiple rules.
+
+### Problems to address:
+- The report should unambiguously point to the rule.
+- Errors and execution time reporting for analyzers.
+- Preventing clashes of identification within a single build: Clashes with well-known rules/analyzers.
+- Possibility to configure the rule.
+- Categorization of the rules/analyzers.
+
+## Proposal
+
+### Built-in analyzers
+Every built-in analyzer will have the friendly name: `BuildCheck.{FriendlyName}`.
+- Regular expression for the name: `^BuildCheck.[A-Z]{1,}[a-zA-Z0-9_]{0,}$`
+
+Each Rule that is shipped inbox will contain the RuleId as an identifier of the rule for this analyzer. 
+- The rule id format is as follows: `^BC[A-Za-z_.]{0,}[0-9]{1,}$`.
+
+#### Example of a built-in analyzer:
+- Name: `BuildCheck.SharedOutputPath`
+- RuleId: `BC0101` or `BC.AdditionalInfo0101` or `BC.Prefix.Test0123`
+
+### Custom analyzers
+Custom analyzer will have the friendly name: `{NameOfTheAnalyzer}` with defined format: 
+- `^[A-Z]{1,}[a-zA-Z_]{1,}$`
+- should not start with `BuildCheck.` this is built-in prefix for built-in.
+
+Each Custom Analyzer Rule will have the rule id format as follows:
+- `^[A-Z]{1}[A-Za-z]{0,}[0-9]{1,}$`. 
+- should not start from `BC.` this is reserved prefix for built-in rules.
+
+#### Example of a custom analyzer:
+- Name: `SharedOutputPath`, `SharedOutputPath` 
+- RuleId: `SOMEPREFIX123`
+
+Any registered analyzers that don't follow the pattern (built-in and custom) will raise an exception and fail the build.
+
+The identification of the rule consists of two components: the Friendlyname and the RuleId.
+
+#### Examples 
+- Built-in
+    - `BuildCheck.SharedOutputPath.BC0001`
+    - `BuildCheck.SharedOutputPath.BC0002`
+    - `BuildCheck.PropertyAssignedIncorrectly.BC0002`
+- Custom
+    - `SharedOutputPath.ID0001`
+    - `SharedOutputPath.ID0002`
+    - `PropertyAssignedIncorrectly.ID0002`
+
+#### Example of the output:
+```
+...
+Determining projects to restore...
+MSBUILD : error : BuildCheck.SharedOutputPath.BC0002: Projects FooBar-Copy.csproj and FooBar.csproj have onflicting output paths: C:\projects\msbuild\playground\buildcheck\bin\Debug\net8.0\.
+MSBUILD : error : BuildCheck.SharedOutputPath.BC0002: Projects FooBar-Copy.csproj and FooBar.csproj have onflicting output paths: C:\projects\msbuild\playground\buildcheck\obj\Debug\net8.0\.
+Restore:
+...
+```
+
+### Categorization
```

Comment:
The categories section feels a little bit out of place right now. I had a bit of trouble understanding how it fit with the analyzer identification topic. I think expanding it a bit would be nice.